### PR TITLE
M3 Button padding adjustments

### LINKED ISSUE #118182: M3 Buttons have incorrect padding values from spec
Raised internally via b/263865808.

According to spec (https://m3.material.io/components/buttons/specs), `Outlined`, `Elevated` and `Filled` buttons should have `horizontal: 24` padding with no icon, and `(left: 16, right 24)` padding with an icon. 

image

`Text` buttons are slightly different and should have `symmetric(horizontal: 12)` padding with no icon, and `(left: 12, right 16)` padding with an icon:

image

However, in the current implementation [`OutlinedButton`](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/material/outlined_button.dart), [`ElevatedButton`](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/material/elevated_button.dart#L394) and [`FilledButton`](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/material/filled_button.dart#L368), padding is set to `horizontal: 16` with no icon, and `left:12, right:16` with an icon. 

[`TextButtons`](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/material/text_button.dart) have `all: 8` with no icon or `left:12, right: 16)` with an icon.

These values are correct for a Material 2 specification (https://m2.material.io/components/buttons#specs), so I think they just need to be updated for M3 values.

### PR DESCRIPTION
**Pull request by esouthren**
Fixes https://github.com/flutter/flutter/issues/118182.

Updates M3 padding values for standard and icon `Outlined`, `Filled`, `Elevated` and `Text` `Buttons`.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/lib/src/material/text_button.dart [hunk 1]
```diff
@@ -378,8 +378,9 @@ class TextButton extends ButtonStyleButton {
 }
 
 EdgeInsetsGeometry _scaledPadding(BuildContext context) {
+  final bool useMaterial3 = Theme.of(context).useMaterial3;
   return ButtonStyleButton.scaledPadding(
-    const EdgeInsets.all(8),
+    useMaterial3 ?  const EdgeInsets.symmetric(horizontal: 12,vertical: 8) :  const EdgeInsets.all(8),
```
Comment (guidezpl) [inline packages/flutter/lib/src/material/text_button.dart:383]:
```suggestion
    useMaterial3 ? const EdgeInsets.symmetric(horizontal: 12, vertical: 8) :  const EdgeInsets.all(8),
```

File: packages/flutter/lib/src/material/text_button.dart [hunk 2]
```diff
@@ -270,7 +270,7 @@ class TextButton extends ButtonStyleButton {
   /// * `shadowColor` - Theme.shadowColor
   /// * `elevation` - 0
   /// * `padding`
-  ///   * `textScaleFactor <= 1` - all(8)
+  ///   * `textScaleFactor <= 1` - (horizontal(12),vertical(8))
```
Comment (guidezpl) [inline packages/flutter/lib/src/material/text_button.dart:273]:
```suggestion
  ///   * `textScaleFactor <= 1` - (horizontal(12), vertical(8))
```

File: packages/flutter/lib/src/material/text_button.dart [hunk 3]
```diff
@@ -320,7 +320,7 @@ class TextButton extends ButtonStyleButton {
   /// * `surfaceTintColor` - null
   /// * `elevation` - 0
   /// * `padding`
-  ///   * `textScaleFactor <= 1` - all(8)
+  ///   * `textScaleFactor <= 1` - lerp(horizontal(8), horizontal(4))
```
Comment (guidezpl) [inline packages/flutter/lib/src/material/text_button.dart:323]: this doesn't seem to match image

Comment (guidezpl) [reply packages/flutter/lib/src/material/text_button.dart:323]: not a huge fan of encoding spec in comments anyway so don't worry about reproducing every last detail

File: packages/flutter/lib/src/material/text_button.dart [hunk 4]
```diff
@@ -491,8 +492,9 @@ class _TextButtonWithIcon extends TextButton {
 
   @override
   ButtonStyle defaultStyleOf(BuildContext context) {
+    final bool useMaterial3 = Theme.of(context).useMaterial3;
     final EdgeInsetsGeometry scaledPadding = ButtonStyleButton.scaledPadding(
-      const EdgeInsets.all(8),
+      useMaterial3 ? const EdgeInsets.only(top: 8, bottom: 8, left: 12, right: 16) : const EdgeInsets.all(8),
```
Comment (guidezpl) [inline packages/flutter/lib/src/material/text_button.dart:497]: suggestion: `fromSTEB`

File: packages/flutter/lib/src/material/outlined_button.dart [hunk 5]
```diff
@@ -287,10 +287,10 @@ class OutlinedButton extends ButtonStyleButton {
   /// * `surfaceTintColor` - null
   /// * `elevation` - 0
   /// * `padding`
-  ///   * `textScaleFactor <= 1` - horizontal(16)
-  ///   * `1 < textScaleFactor <= 2` - lerp(horizontal(16), horizontal(8))
-  ///   * `2 < textScaleFactor <= 3` - lerp(horizontal(8), horizontal(4))
-  ///   * `3 < textScaleFactor` - horizontal(4)
+  ///   * `textScaleFactor <= 1` - horizontal(24)
```
Comment (guidezpl) [inline packages/flutter/lib/src/material/outlined_button.dart:290]: maybe note "less with an icon"?

**Review by guidezpl — CHANGES_REQUESTED** [review]
Would be great to run the sample demo with this PR to visually verify everything

File: packages/flutter/lib/src/material/text_button.dart (hunk 3 above)
Comment (esouthren) [reply packages/flutter/lib/src/material/text_button.dart:323]: done

File: packages/flutter/lib/src/material/text_button.dart (hunk 4 above)
Comment (esouthren) [reply packages/flutter/lib/src/material/text_button.dart:497]: Ah, better - thank you

File: packages/flutter/lib/src/material/outlined_button.dart (hunk 5 above)
Comment (esouthren) [reply packages/flutter/lib/src/material/outlined_button.dart:290]: Good addition - added to all buttons.

**Comment by esouthren** [comment]
> Would be great to run the sample demo with this PR to visually verify everything

Sure - ran against the new M3 app:
image

The most noticable difference is between the Text button icon padding and the others, as it's now slightly different (off centre compared to the [current app](https://flutter-experimental-m3-demo.web.app/#/)

**Review by guidezpl — APPROVED** [review]

**Comment by godofredoc** [comment]
Reverting as it broke framework tests https://ci.chromium.org/p/flutter/builders/prod/Linux%20framework_tests_libraries/9116

**Comment by bernaferrari** [comment]
Could this be fixed since M3 is breaking anyway? https://github.com/flutter/flutter/issues/85729

**Comment by esouthren** [comment]
> Reverting as it broke framework tests https://ci.chromium.org/p/flutter/builders/prod/Linux%20framework_tests_libraries/9116

Thanks - will take a look and try to reland.

**Pull request merged**